### ISSUE TITLE
Freezing layers of the model

### ISSUE THREAD

**Issue by ChristianMeli** [bcmi/SLBR-Visible-Watermark-Removal#34]
Hello.

I would like to experiment with freezing some weights of the model. **I want to keep the core knowledge learned but still give it room to learn. What weights should be frozen?**

If this were a simple Sequential model or a large pretrained model to be used on a downstream task, I would just freeze all the weights except the ones from the head (if the task was a classification, for instance). Is there an equivalent here? 

This is the output I get from summary function from [torch-summary](https://pypi.org/project/torch-summary/) library, if it helps to see it fast.

```
===========================================================================
Layer (type:depth-idx)                             Param #
===========================================================================
├─CoarseEncoder: 1-1                               --
|    └─ModuleList: 2-1                             --
|    |    └─DownConv: 3-1                          28,896
|    |    └─DownConv: 3-2                          129,792
|    |    └─DownConv: 3-3                          517,632
├─SharedBottleNeck: 1-2                            --
|    └─ModuleList: 2-2                             --
|    |    └─DownConv: 3-4                          2,065,408
|    |    └─DownConv: 3-5                          8,259,584
|    └─ModuleList: 2-3                             --
|    |    └─UpConv: 3-6                            4,130,048
|    └─ModuleList: 2-4                             --
|    |    └─ECABlock: 3-7                          3
|    └─ModuleList: 2-5                             --
|    |    └─ECABlock: 3-8                          3
├─CoarseDecoder: 1-3                               --
|    └─Conv2d: 2-6                                 99
|    └─ModuleList: 2-7                             --
|    |    └─MBEBlock: 3-9                          970,147
|    |    └─MBEBlock: 3-10                         243,923
|    |    └─MBEBlock: 3-11                         61,675
|    └─ModuleList: 2-8                             --
|    |    └─ECABlock: 3-12                         3
|    |    └─ECABlock: 3-13                         3
|    |    └─ECABlock: 3-14                         3
|    └─ModuleList: 2-9                             --
|    |    └─SMRBlock: 3-15                         1,135,782
|    |    └─SMRBlock: 3-16                         284,758
|    |    └─SMRBlock: 3-17                         71,598
|    └─ModuleList: 2-10                            --
|    |    └─ECABlock: 3-18                         3
|    |    └─ECABlock: 3-19                         3
|    |    └─ECABlock: 3-20                         3
├─Refinement: 1-4                                  --
|    └─Sequential: 2-11                            --
|    |    └─Conv2d: 3-21                           1,184
|    |    └─InstanceNorm2d: 3-22                   --
|    |    └─LeakyReLU: 3-23                        --
|    └─ResDownNew: 2-12                            --
|    |    └─DownConv: 3-24                         36,992
|    └─ResDownNew: 2-13                            --
|    |    └─DownConv: 3-25                         129,280
|    └─ResDownNew: 2-14                            --
|    |    └─DownConv: 3-26                         516,608
|    └─Sequential: 2-15                            --
|    |    └─Conv2d: 3-27                           1,056
|    └─Sequential: 2-16                            --
|    |    └─Conv2d: 3-28                           2,080
|    |    └─LeakyReLU: 3-29                        --
|    |    └─Conv2d: 3-30                           9,248
|    |    └─LeakyReLU: 3-31                        --
|    └─Sequential: 2-17                            --
|    |    └─Conv2d: 3-32                           8,256
|    |    └─LeakyReLU: 3-33                        --
|    |    └─Conv2d: 3-34                           36,928
|    |    └─LeakyReLU: 3-35                        --
|    └─ModuleList: 2-18                            --
|    |    └─CFFBlock: 3-36                         895,104
|    |    └─CFFBlock: 3-37                         895,104
|    |    └─CFFBlock: 3-38                         895,104
|    └─Sequential: 2-19                            --
|    |    └─Conv2d: 3-39                           64,544
|    |    └─InstanceNorm2d: 3-40                   --
|    |    └─LeakyReLU: 3-41                        --
|    |    └─Conv2d: 3-42                           99
===========================================================================
Total params: 21,390,953
Trainable params: 21,390,953
Non-trainable params: 0
```
 
I've tried just finetunning the model without freezing, of course, but I've seen it degrading in some examples. As in, the checkpoint you provide removes the watermark but after training I see some parts of the watermark coming back. 

Thanks for your help :)

**Comment by jimleungjing** [comment]
1. you can finetune the model by small learning rate, e.g. [1e-5, 1e-6]. 
2. You can finetune the model start from the decoder to encoder, and layer by layer. For example, finetune the decoder shallow layers while freezing the other parts. 
3.  You can try the most popular finetune model LoRA to our pretrained model.
